proposals/multiple_user_script_worlds.md

```diff
@@ -0,0 +1,285 @@
+are injected, since all user scripts will run either in the context of the
+main world or in a single user script world for the extension. This API would
+allow user script managers to instead have multiple user script worlds, each
+isolated from one another, to reduce the likelihood of user scripts negatively
+affecting one another through collisions.
+
+#### Use Cases
+
+The primary use case is user script managers that handle multiple, independent
+script injections in a single page.
+
+### Known Consumers
+
+We anticipate user script managers such as Tampermonkey, Violentmonkey, and
+others to use this new capability.
+
+## Specification
+
+### Schema
+
+This change would modify existing schemas.
+
+Relevant methods and types:
+
+```
+export namespace userScripts {
+  export interface WorldProperties {
+    /**
+     * Specifies the world csp. The default is the `` `ISOLATED` `` world csp.
+     */
+    csp?: string;
+
+    /**
+     * Specifies whether messaging APIs are exposed. The default is `false`.
+     */
+    messaging?: boolean;
+  }
+
+  /**
+   * Properties for a registered user script.
+   */
+  export interface RegisteredUserScript {
+    /**
+     * If true, it will inject into all frames, even if the frame is not the
+     * top-most frame in the tab. Each frame is checked independently for URL
+     * requirements; it will not inject into child frames if the URL
+     * requirements are not met. Defaults to false, meaning that only the top
+     * frame is matched.
+     */
+    allFrames?: boolean;
+    /**
+     * Specifies wildcard patterns for pages this user script will NOT be
+     * injected into.
+     */
+    excludeGlobs?: string[];
+    /**
+     * Excludes pages that this user script would otherwise be injected into.
+     * See Match Patterns for more details on the syntax of these strings.
+     */
+    excludeMatches?: string[];
+    /**
+     * The ID of the user script specified in the API call. This property
+     * must not start with a '_' as it's reserved as a prefix for generated
+     * script IDs.
+     */
+    id: string;
+    /**
+     * Specifies wildcard patterns for pages this user script will be
+     * injected into.
+     */
+    includeGlobs?: string[];
+    /**
+     * The list of ScriptSource objects defining sources of scripts to be
+     * injected into matching pages.
+     */
+    js: ScriptSource[];
+    /**
+     * Specifies which pages this user script will be injected into. See
+     * Match Patterns for more details on the syntax of these strings. This
+     * property must be specified for ${ref:register}.
+     */
+    matches?: string[];
+    /**
+     * Specifies when JavaScript files are injected into the web page. The
+     * preferred and default value is document_idle
+     */
+    runAt?: RunAt;
+    /**
+     * The JavaScript execution environment to run the script in. The default
+     * is `USER_SCRIPT`
+     */
+    world?: ExecutionWorld;
```

Comment:
Thanks for raising this.  I agree with aligning on a uniform standard, but would prefer using all caps for the reasons I highlighted in the response there.

I'd like to avoid blocking this issue on the resolution of that (it seems like it is very likely to have lots of impassioned bikeshed discussions : )); for the purpose of this proposal, I'd propose we keep the all caps since that aligns with the rest of the user script API, and we can establish what to do for other cases as part of issue #563  (since we'll need to address other APIs anyway).  Does that sound reasonable to you?